Predictions: Allow choosing a target

janezd opened the pull request:
##### Issue

Fixes #5743.

##### Description of changes

It sure looks funny, especially for numeric targets, but.

I didn't want to put the combo to the main area in order to save space; in Test and Score we have plenty space on the right, here it's precious. I have put the Restore Order button and the list view for selecting probabilities at the top, and the combo for choosing the target at the bottom (with rubber in between), so they are close to their respective tables.

<img width="778" alt="Screenshot 2022-01-14 at 23 00 12" src="https://user-images.githubusercontent.com/2387315/149590979-f00c0fbe-911d-43a4-965b-b92d2704f1ec.png">

This is quite nice for discrete outcomes; less so for numeric.

<img width="778" alt="Screenshot 2022-01-14 at 23 00 22" src="https://user-images.githubusercontent.com/2387315/149591028-a2689a62-8942-4cd8-8c15-ac72297487c3.png">

I've no idea what to do here except for perhaps having a different layout in this case: just put the checkbox and the button between the tables (in that order, with rubber between), and hide the control area completely.

##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

ajdapretnar commented:
What if:
- probabilities are always shown
- model scores are always shown
- target class can be selected above performance scores if problem is categorical
Ok, I've come so far but I've no idea where to put "Restore original order". 🙈

janezd commented:
What about this?

I replaced the listview with a combo and moved everything to main area. This not only gets rid of the empty control area and makes everything nicer, it also simplifies the widget and its code.

The listview contained all classes that appeared in the data or in any of the models. So the user was able to choose any combination of classes, including, say, some classes that only appeared in some models. I doubt anybody used this, so the list was a nuissance. Why would anybody be interested in setosa and virginica, but not versicolor?

Now the combo offers only sensible options: no probabilities, probabilities that appear in the data, probabilities that are predicted by models (which may differ from those in the data) and probabilities of individual classes that appear in the data. No combinations.

The widget did not work with contexts: it couldn't, because the context would be an ordered combinations of all classes appearing in any model (because this is what list view offered). Now, the context is simply the list of classes in data.

<img width="920" alt="Screenshot 2022-02-11 at 18 48 29" src="https://user-images.githubusercontent.com/2387315/153642989-a619ce08-0b67-40ed-ae51-be70fb132572.png">

<img width="870" alt="Screenshot 2022-02-11 at 18 49 06" src="https://user-images.githubusercontent.com/2387315/153643014-50e189e8-af5e-4549-91a5-606cb016d838.png">

markotoplak commented:
@janezd, I like the concept, but your new approach trades valuable vertical space for (cheap an collapsible) horizontal. Substantially decreasing margins around components would likely make me stop complaining.

janezd commented:
@markotoplak, you're right. Is this better?

For numeric target, I could move Restore Original Order below the table to eliminate the components above, but this could be confusing.

<img width="797" alt="Screenshot 2022-02-11 at 19 34 53" src="https://user-images.githubusercontent.com/2387315/153649720-ffa251b2-c481-4d33-a225-08af3ce96bb8.png">

markotoplak commented:
Looks good! I do not know about the Qt quirks about this one, but the upper control pane still has larger margins than the lower one.

I agree that moving the button could be confusing.

janezd commented:
It turns out that it was not a Qt's quirk: `gui.button` adds some space.

```python
    if is_macstyle():
        btnpaddingbox = vBox(widget, margin=0, spacing=0)
        separator(btnpaddingbox, 0, 4)  # lines up with a WA_LayoutUsesWidgetRect checkbox
        button.outer_box = btnpaddingbox
```

Ready for review.

markotoplak commented:
Hahaha, so all our buttons are adding padding that can not be disabled just so that they align with potential checkboxes? Who merged that? Wait, it was me. :)

borondics commented:
Hm... 🤪

VesnaT commented on Orange/widgets/evaluate/owpredictions.py:
```diff
@@ -682,10 +801,18 @@ def merge_data_with_predictions():
+                            "and are known to the model"
+                else:
+                    class_idx = self.shown_probs - len(self.PROB_OPTS)
+                    text += f"'{self.class_var.values[class_idx]}'"
```
The `target_class` value is missing in the report.

VesnaT replied:
Are you sure you want to display it for regression as well?

janezd replied:
Sure not.

janezd commented on Orange/widgets/evaluate/owpredictions.py:
```diff
@@ -76,8 +75,21 @@ class Error(OWWidget.Error):
+                     "Show probabilities for classes in data that are also\n"
+                     "known to the model"
+                     ]
+    NO_PROBS, DATA_PROBS, MODEL_PROBS, BOTH_PROBS = range(4)
```
Fortunately not.

First, the list was composed of all classes that appeared in any of connected predictors. This information is not a part of the `ClassValuesContextHandler` that knows only about the values that appear in the data's class. We can't go from indices in this list to names of values. And even if we could, there wouldn't always be a correct option to choose from.

Second, this never worked anyway because the widget didn't close and open contexts. :)

janezd commented on Orange/widgets/evaluate/owpredictions.py:
```diff
@@ -202,24 +246,43 @@ def insert_predictor(self, index, predictor: Model):
+            self.shown_probs = self.DATA_PROBS
+            self.target_class = self.TARGET_AVERAGE
+        else:
+            self.shown_probs = self.NO_PROBS
```
Huh, I could but this would complicate the code and the widget.

- In terms of UX, allowing `MODEL_PROBS` when the data class is continuous but some models are categorical would require showing the combo, which would then have two options, `NO_PROBS` and `MODEL_PROBS`. When all models are regressors, the combo would be hidden...
- In terms of the code, I can now simply expect that `shown_probs` is `NO_PROBS` in case of numeric data, which simplifies some conditions, as I remember.

Though I could change this, I'd prefer not to. This situation is too strange; perhaps we could even ignore models of wrong kind and show a warning.

VesnaT replied:
Ok, leave it as is. But be aware that, in e.g. six months, one may report an issue: Predictions widget is missing probabilities. :)

janezd replied:
We'll tell him that he should not mix classifiers and regressors. :)

janezd commented on Orange/widgets/evaluate/owpredictions.py:
```diff
@@ -233,6 +296,16 @@ def handleNewSignals(self):
+        # because one commit button has only one dirty flag, and unnecessarily
+        # making an unconditional commit af a small table is better than
+        # conditionally commit all predictions.
+        self._commit_evaluation_results()
```
Good catch. I assumed that the output depends on what is shown, but it does not.

I now fixed that. The output table now contains the same data as shown in the view. And thus depend on the combo. This is relevant only for classification, so combo can still be hidden for regression.

VesnaT replied:
I was actually referring to the Evaluation results output.
I don't see how those depend on the target combo box.

janezd replied:
Oh, you're right. I removed the comment and the function call.

janezd commented on Orange/widgets/evaluate/owpredictions.py:
```diff
@@ -86,22 +98,43 @@ def __init__(self):
+        button.clicked.connect(self._reset_order)
+        button.setToolTip("Show rows in the original order")
+        button.setSizePolicy(QSizePolicy.Fixed, QSizePolicy.Fixed)
+        predopts.layout().addWidget(self.reset_button)
```
Fixed.

janezd commented on Orange/widgets/evaluate/owpredictions.py:
```diff
@@ -202,24 +246,43 @@ def insert_predictor(self, index, predictor: Model):
+            widget.setVisible(self.is_discrete_class)
+
+        for widget in self._target_controls:
+            widget.setVisible(self.is_discrete_class and self.show_scores)
```
Fixed -- hiding entire box when there is no target.

janezd commented:
Here's a workflow to explore the behaviour of the output.

[test-predictions.ows.zip](https://github.com/biolab/orange3/files/8101786/test-predictions.ows.zip)
.

VesnaT commented:
I've found another minor issue, but it might had been there before:

1. File (Iris) -> Logistic Regression -> Predictions
2. Predictions (Show probabilities for `Iris-setosa`) -> Sort descending by Logistic Regression column 
3. Predictions (Show probabilities for `Iris-versicolor`) -> Logistic Regression column should be sorted descending (due to sort indicator), but it is not

janezd commented:
> I've found another minor issue, but it might had been there before: (...) Logistic Regression column should be sorted descending (due to sort indicator), but it is not

Quite probably. The argument for order was missing when sorting after changing indices: https://github.com/biolab/orange3/pull/5790/files#diff-dd77f983a656d8fa3f453496f08aad0ae7e8c5a57e48406d5520836c76ce1690R1040.

The changes were so minor (removed call to resend signals; `if` that checks for discrete target in report; added argument) that I just amended the existing commit. Ready for rerereview.

VesnaT commented:
> Quite probably. The argument for order was missing when sorting after changing indices: https://github.com/biolab/orange3/pull/5790/files#diff-dd77f983a656d8fa3f453496f08aad0ae7e8c5a57e48406d5520836c76ce1690R1040.

The sort indicator is now shown on the first column by default, which is not ok.

janezd commented:
> The sort indicator is now shown on the first column by default, which is not ok.

True, but without "now". This misbehaviour is old.

This should fix it: https://github.com/biolab/orange3/pull/5790/files#diff-dd77f983a656d8fa3f453496f08aad0ae7e8c5a57e48406d5520836c76ce1690R523.
